### PR TITLE
Add TruyenGG

### LINKED ISSUE #5461: [VI] TruyenGG
### Source name

TuyenGG

### Source link

https://truyengg.com/

### Source language

VI-Vietnam

### Other details

NSFW = Yes

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by dejavui**
Checklist:
Closes #5461
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/vi/truyengg/build.gradle [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ext {
+    extName = 'TruyenGG'
+    extClass = '.TruyenGG'
+    extVersionCode = 0
```
Comment (vetleledaal) [inline src/vi/truyengg/build.gradle]: CI is failing here
```suggestion
    extVersionCode = 1
```

File: src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        .retryOnConnectionFailure(true)
+        .followRedirects(true)
+        .rateLimit(2)
+        .build()
+
+    override fun headersBuilder(): Headers.Builder =
+        super.headersBuilder().add("Referer", "$baseUrl/")
+
+    override fun chapterFromElement(element: Element): SChapter = SChapter.create().apply {
+        setUrlWithoutDomain(element.select("a").attr("href"))
```
Comment (vetleledaal) [inline src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt]: Use `.selectFirst()` when applicable. Comment applies throughout this file. Using `.select()` is less performant (not that important) and will select all elements which may lead to unintended behavior.

**Review by vetleledaal — CHANGES_REQUESTED** [review]

File: src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+    override val lang = "vi"
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .connectTimeout(30, TimeUnit.SECONDS)
+        .readTimeout(30, TimeUnit.SECONDS)
+        .retryOnConnectionFailure(true)
+        .followRedirects(true)
+        .rateLimit(1)
+        .build()
```
Comment (AwkwardPeak7) [inline src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt]:
```suggestion
        .retryOnConnectionFailure(true)
        .rateLimit(1)
        .build()
```
the others are default behavior on tachi/mihon.

File: src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    override val baseUrl by lazy { getPrefBaseUrl() }
+
+    private val defaultBaseUrl = "https://truyengg.com"
+
+    override val lang = "vi"
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .connectTimeout(30, TimeUnit.SECONDS)
+        .readTimeout(30, TimeUnit.SECONDS)
+        .retryOnConnectionFailure(true)
```
Comment (AwkwardPeak7) [inline src/vi/truyengg/src/eu/kanade/tachiyomi/extension/vi/truyengg/TruyenGG.kt]: why?

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**